# TP n°6 : Fonctions 1/2



___________________________________________________________________________
## Fonction de calcul du volume du ballon de rugby



Consulter l'article de wikipedia donnant le volume d'un [ellipsoïde de révolution](https://fr.wikipedia.org/wiki/Ellipso%C3%AFde_de_r%C3%A9volution#Volume_int.C3.A9rieur).

Ecrire une fonction `volumeRugby(H, L)` qui détermine le volume d'un ballon de rugby de hauteur `H` et de largeur `L`. On écrira explicitement la formule utilisée par Wikipedia, avec ses variables et après avoir fait le lien avec les variables `H` et `L`.

On importera `pi` avec le module math.

Appliquer à un ballon standard de tailles $H = 29 \text{cm}$ et $L=  19 \text{cm}$, de volume environ 5,5 litres.



In [12]:
from math import pi
print(int(input("h (en cm) = "))*int(input("l (en cm) = "))**2*pi*4/(24*1000))

h (en cm) = 29
l (en cm) = 19
5.481555581738592


___________________________________________________________________________
## Implémenter la fonction `max`



Écrire une fonction `maxi` qui renvoie le maximum de deux nombres `a` et `b` passés en paramètre.



In [4]:
def maxi(a, b):
    if a > b: return a
    return b
print(maxi(3, 5))

5


___________________________________________________________________________
## Implémenter la fonction *valeur absolue*



Écrire une fonction `val_abs` qui retourne la valeur absolue d'un nombre `x` passé en paramètre.



In [15]:
def val_abs(x):
    return int((x**2)**.5)
print(val_abs(-6))

6


___________________________________________________________________________
## Fonction pour année bissextile



Écrire une fonction `estBissextile` qui prend en paramètre un entier `a` et retourne `True` si l'année `a` est bissextile (`False` sinon). Pour rappel, une année `a` est bissextile si le booléen suivant

`(a%4==0 and a%100!=0) or a%400==0`

vaut `True`.



In [14]:
def estBissextile(a):
    return a % 4 == 0 and a % 100 != 0 or a % 400 == 0
print(estBissextile(1956))

True


___________________________________________________________________________
## Fonctions Celsius <-> Fahrenheit



1) La formule de conversion de degrés Fahrenheit vers des degrés Celsius est

$$c=\frac{5}{9}(f-32) $$



a) Écrire une fonction `f2c(f)` qui prend en paramètre une température $f$ (en degrés Fahrenheit) et qui renvoie sa conversion $c$ (en degrés Celsius).



In [ ]:
def f2c(f):
    return 5/9*(f - 32)

b) Combien font $451^\circ F$ et $0 ^\circ F$ en $^\circ C$ ?



In [17]:
def f2c(f):
    return 5/9*(f - 32)
print(f2c(451), f2c(0), sep = "\n")

232.7777777777778
-17.77777777777778


2) La formule de conversion de degrés Celsius vers des degrés Fahrenheit est  $$f=\frac{9}{5}c+32 $$



a) Écrire une fonction `c2f(c)` qui prend en paramètre une température $c$ (en degrés Celsius) et qui renvoie sa conversion $f$ (en degrés Fahrenheit)



In [ ]:
def c2f(c):
    return 9/5*c + 32

b) Combien font $42^\circ C$ et $-273,15 ^\circ C$ en $^\circ F$ ?



In [19]:
def c2f(c):
    return 9/5*c + 32
print(c2f(42), c2f(-273.15), sep = "\n")

107.60000000000001
-459.66999999999996


3) Pour trois valeurs `x` qui vous choisirez librement, vérifiez que `c2f(f2c(x))` et `f2c(c2f(x))` sont proches de `x`.



In [24]:
def f2c(f):
    return 5/9*(f - 32)
def c2f(c):
    return 9/5*c + 32
print(c2f(f2c(13)), f2c(c2f(13)), sep = "\n")

13.0
13.000000000000004


___________________________________________________________________________
## Calcul d'une durée



1) Écrire une fonction `hms_to_s(h, m, s)` qui retourne en secondes une durée exprimée en heures, minutes et secondes. Vérifier que 1h30m = 5400s et 3h20m15s = 12015s.



In [21]:
def hms_to_s(h, m, s):
    return 3600*h + 60*m + s
print(hms_to_s(2, 30, 55))

9055


2) Écrire une fonction `s_to_hms(sec)` qui prend en paramètre une durée exprimée en secondes et la retourne exprimée en heures, minutes et secondes sous forme de liste [h, m, s]. Vérifier que `s_to_hms(5400)` retourne la liste `[1, 30, 0]`, et que `s_to_hms(12015)=[3, 20, 15]`.



In [29]:
def s_to_hms(s):
    return [s//3600, (s % 3600)//60, s % 60]
print(s_to_hms(5400))
print(s_to_hms(12015))

[1, 30, 0]
[3, 20, 15]


3) Écrire une fonction `afficher_temps` capable d'afficher une liste [h, m, s] sous la forme `h heures m minutes et s secondes`. Par exemple, `afficher_temps([2, 42, 16])` affiche `2 heures 42 minutes et 16 secondes`.



In [38]:
def afficher_temps(l):
    return str(abs(l[0])) + " heures " + str(abs(l[1])) + " minutes et " + str(abs(l[2])) + " secondes."
print(afficher_temps([2, 42, 16]))

2 heures 42 minutes et 16 secondes.


4) Sans définir de nouvelle fonction, écrire un code, tenant de préférence sur une seule ligne, qui affiche le temps écoulé comme à la question précédente entre 13h58m et 15h31m30s.



In [39]:
def afficher_temps(l):
    return str(abs(l[0])) + " heures " + str(abs(l[1])) + " minutes et " + str(abs(l[2])) + " secondes."
print(afficher_temps([15 - 13, 31 - 58, 30]))

2 heures 27 minutes et 30 secondes.


___________________________________________________________________________
## Équation du second degré (version fonction)



On rappelle ci-dessous un code de résolution et d'affichage des solutions d'une équation du second degré $ax^2+bx+c=0$ mais n'utilisant ni fonction ni liste :



In [ ]:
a = 6
b = -5
c = 1
delta = b*b -4*a*c
if delta >0:
    print("2 solutions distinctes :")
    print("x1 =", (-b - delta**0.5)/(2.*a))
    print("x2 =", (-b + delta**0.5)/(2.*a))
elif delta == 0:
    print("une seule solution")
    print("x =", (-b)/(2.*a))
else:
    print("Aucune solution")


Vous devez adapter ce code pour les questions ci-dessous.



1) Ecrire une fonction `resoudre_2nd_degre(a, b, c)` qui résout dans l'ensemble des nombres réels l'équation du second degré. La fonction prend en paramètre $\mathtt{a, b, c}$ les coefficients supposés entiers de l'équation $\mathtt{ax^2+bx+c=0}$ et renvoie une **liste** de solutions, éventuellement vide.



In [41]:
def resoudre_2nd_degre(a, b, c):
    delta = b*b -4*a*c
    if delta > 0:
        x1, x2 = (-b - delta**0.5)/(2.*a), (-b + delta**0.5)/(2.*a)
        return [x1, x2]
    elif delta == 0:
        x = (-b)/(2.*a)
        return [x]
    else: return []
a, b, c = 6, -5, 1   
print(resoudre_2nd_degre(a, b, c))

[0.3333333333333333, 0.5]


2) Ecrire une fonction d'affichage `afficher_solutions(a, b, c)` des solutions. La fonction affichage doit utiliser la fonction précédente.

Voici un exemple de comportements du code à écrire :

    a = 6 b = -5 c = 1 :
    
    2 solutions distinctes :
    x1 = 0.3333333333333333
    x2 = 0.5
    ----------------------------------------------
    a = 4 b = -12 c = 9 :
    
    une seule solution
    x = 1.5
    ----------------------------------------------
    a = 6 b = 7 c = 7 :
          
    Aucune solution
    ----------------------------------------------


In [5]:
def resoudre_2nd_degre(a, b, c):
    delta = b*b -4*a*c
    if delta > 0:
        x1, x2 = (-b - delta**0.5)/(2.*a), (-b + delta**0.5)/(2.*a)
        return [x1, x2]
    elif delta == 0:
        x = (-b)/(2.*a)
        return [x]
    else: return []

def afficher_solutions(a, b, c):
    l = resoudre_2nd_degre(a, b, c)
    print("a", "=", a, ";", "b", "=", b, ";", "c", "=", c, ":", end = "\n\n")
    nbS = len(l)
    if nbS == 0: print("Aucune solution.")
    elif nbS == 1: print("Une seule solution :\nx =", l[0])
    else: print("2 solutions distinctes :\nx1 =", l[0], "\nx2 =", l[1])
    print("\n" + "-"*42 + "\n")

afficher_solutions(6, -5, 1)
afficher_solutions(4, -12, 9)
afficher_solutions(6, 7, 7)

a = 6 ; b = -5 ; c = 1 :

2 solutions distinctes :
x1 = 0.3333333333333333 
x2 = 0.5

------------------------------------------

a = 4 ; b = -12 ; c = 9 :

Une seule solution :
x = 1.5

------------------------------------------

a = 6 ; b = 7 ; c = 7 :

Aucune solution.

------------------------------------------



___________________________________________________________________________
## Fonction binomiale



1) Écrire une fonction `factorielle` qui renvoie la valeur factorielle de `n` où `n` étant passé en paramètre. Par exemple $1!=1$ et $5!=120$. On convient que $0!=1$. On écrira une version de la fonction factorielle utilisant une boucle `for`.



In [4]:
def factorielle(n):
    fac = 1
    for cpt in range(2, n + 1):
        fac *= cpt
    return fac

for _ in range(3):
    print(factorielle(int(input())), end = "\n\n")

1
1

5
120

0
1



2) On rappelle que si $n$ et $p$ sont des entiers positifs ou nuls, le coefficient binomial est défini par

$$\binom{n}{p}=\begin{cases}\displaystyle\frac{n!}{p!(n-p)!}& \text{si } 0\leq p\leq n\\0& \text{sinon}\end{cases}$$

Ecrire une fonction `binomial(n, p)` qui renvoie le coefficient binomial ci-dessus.



In [ ]:
def factorielle(n):
    fac = 1
    for cpt in range(2, n + 1):
        fac *= cpt
    return fac

def binomial(n, p):
    if 0 <= p <= n: return factorielle(n)//(factorielle(p)*factorielle(n - p))
    return 0

3) Ecrire une fonction `pascal` qui vérifie l'identité  de Pascal : $\binom{n}{p}+\binom{n}{p+1}=\binom{n+1}{p+1}$ valable pour tous entiers $n, p\geq 0$. Vérifier l'identité pour tous les entiers $n$ et $p$ entre 0 et 100.



In [8]:
def factorielle(n):
    fac = 1
    for cpt in range(2, n + 1):
        fac *= cpt
    return fac

def binomial(n, p):
    if 0 <= p <= n: return factorielle(n)//(factorielle(p)*factorielle(n - p))
    return 0

def pascal(n, p):
    return binomial(n, p) + binomial(n, p + 1) == binomial(n + 1, p + 1)

for entN in range(101):
    for entP in range(101):
        if not pascal(entN, entP): print(entN, entP)
        print(pascal(entN, entP))

True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True


True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True


True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True


True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True


True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True
True


4) Ecrire une fonction `maxBinomial(n)` qui renvoie la valeur maximale de tous les coefficients binomiaux  $\binom{n}{0}, \binom{n}{1}, \binom{n}{2}, \cdots , \binom{n}{n}$



In [ ]:
def factorielle(n):
    fac = 1
    for cpt in range(2, n + 1):
        fac *= cpt
    return fac

def binomial(n, p):
    if 0 <= p <= n: return factorielle(n)//(factorielle(p)*factorielle(n - p))
    return 0

def maxBinomial(n):
    return max([binomial(n, p) for p in range(n + 1)])

5) Ecrire une fonction `verifMaxBinomial(n)` qui vérifie que la valeur maximale de tous les coefficients binomiaux $\binom{n}{0}, \binom{n}{1}, \binom{n}{2}, \cdots, \binom{n}{n}$ est $\binom{n}{m}$ où $m$ est le quotient la division entière de $n$ par 2.



In [ ]:
def factorielle(n):
    fac = 1
    for cpt in range(2, n + 1):
        fac *= cpt
    return fac

def binomial(n, p):
    if 0 <= p <= n: return factorielle(n)//(factorielle(p)*factorielle(n - p))
    return 0

def maxBinomial(n):
    return max([binomial(n, p) for p in range(n + 1)])

def verifMaxBinomial(n):
    return maxBinomial(n) == binomial(n, n//2)

___________________________________________________________________________
## Disques creux



1) Écrire une procédure `disque` permettant d'afficher avec Turtle un disque creux dont les coordonnées `x` et `y` du centre et le diamètre `d` sont passés en paramètre. On tracera un `dot` de couleur `white` à l'intérieur d'un `dot` de couleur `black`.



In [16]:
from turtle import *
reset()
ht()

def disque(x, y, d):
    up()
    goto(x, y)
    down()
    dot(d, "black")
    dot(d//2, "white")

disque(100, 100, 100)

2) Dans le programme principal, dessiner une grille de disques ayant la forme de la figure suivante.

![](disques_trou.png)

L'espacement entre les disques correspond au diamètre de chaque disque noir.



In [1]:
from turtle import *
reset()

def disque(x, y, d):
    up()
    goto(x, y)
    down()
    dot(d, "black")
    dot(d//2, "white")

def grille(nb, d):
    start = -(nb - 1)*d//2
    y = -start
    for _ in range(nb):
        x = start
        for _ in range(nb):
            disque(x, y, d)
            x += d
        y -= d

grille(int(input()), int(input()))

4
100
